python/coinbase-agentkit/coinbase_agentkit/wallet_providers/cdp_evm_smart_wallet_provider.py

```diff
@@ -0,0 +1,396 @@
+                else 1.2
+            )
+
+            self._fee_per_gas_multiplier = (
+                max(config.gas.fee_per_gas_multiplier, 1)
+                if config and config.gas and config.gas.fee_per_gas_multiplier is not None
+                else 1
+            )
+
+        except ImportError as e:
+            raise ImportError(
+                "Failed to import cdp. Please install it with 'pip install cdp-sdk'."
+            ) from e
+        except Exception as e:
+            raise ValueError(f"Failed to initialize CDP smart wallet: {e!s}") from e
+
+    def get_client(self) -> CdpClient:
+        """Get a new CDP client instance.
+
+        Returns:
+            Cdp: A new CDP client instance
+
+        """
+        return CdpClient(
+            api_key_id=self._api_key_id,
+            api_key_secret=self._api_key_secret,
+            wallet_secret=self._wallet_secret,
+            debugging=True,
+        )
+
+    def _run_async(self, coroutine):
+        """Run an async coroutine synchronously.
+
+        Args:
+            coroutine: The coroutine to run
+
+        Returns:
+            Any: The result of the coroutine
+
+        """
+        try:
+            loop = asyncio.get_event_loop()
+        except RuntimeError:
+            loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(loop)
+        return loop.run_until_complete(coroutine)
+
+    def get_address(self) -> str:
+        """Get the wallet address.
+
+        Returns:
+            str: The wallet's address as a hex string
+
+        """
+        return self._address
+
+    def get_balance(self) -> Decimal:
+        """Get the wallet balance in native currency.
+
+        Returns:
+            Decimal: The wallet's balance in wei as a Decimal
+
+        """
+        balance = self._web3.eth.get_balance(self.get_address())
+        return Decimal(balance)
+
+    def get_name(self) -> str:
+        """Get the name of the wallet provider.
+
+        Returns:
+            str: The string 'cdp_evm_smart_wallet_provider'
+
+        """
+        return "cdp_evm_smart_wallet_provider"
+
+    def get_network(self) -> Network:
+        """Get the current network.
+
+        Returns:
+            Network: Network object containing protocol family, network ID, and chain ID
+
+        """
+        return self._network
+
+    def native_transfer(self, to: str, value: Decimal) -> str:
+        """Transfer the native asset of the network using a user operation.
+
+        Args:
+            to (str): The destination address to receive the transfer
+            value (Decimal): The amount to transfer in whole units (e.g. 1.5 for 1.5 ETH)
+
+        Returns:
+            str: The transaction hash as a string
+
+        """
+        value_wei = Web3.to_wei(value, "ether")
+        client = self.get_client()
+
+        async def _send_user_operation():
+            async with client as cdp:
+                smart_account = await cdp.evm.get_smart_account(
```

Comment:
any reason you're not also storing smart_account as a class variable?

Reply:
I made an assumption that it would have the same socket issues as storing the CDP client. After testing, this was not true, so I am now storing the smart account directly